File: MPP-casestudy/pgmio.c
/*
 * This file contains C routines for the MPI Casestudy.
 *
 * "pgmread" reads in a PGM picture and can be called as follows:
 *
 *    float buf[M][N];
 *    pgmread("edge.pgm", buf, M, N);
 *
 * "pgmwrite" writes an array as a PGM picture and can be called as follows:
 *
 *    float buf[M][N];
 *    pgmwrite("picture.pgm", buf, M, N);
 *
 * "pgmsize" returns the size of a PGM picture and can be called as follows:
 *
 *    int nx, ny;
 *    pgmsize("edge.pgm", &nx, &ny);
 *
 *  To access these routines, add the following to your program:
 *
 *    #include "pgmio.h"
 *
 *  Note: you MUST link with the maths library -lm to access fabs etc.
 */


#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MAXLINE 128

/*
 *  Routine to get the size of a PGM data file
 *
 *  Note that this assumes a single line comment and no other white space.
 */

void pgmsize(char *filename, int *nx, int *ny)
{ 
  FILE *fp;

  char *cret;
  int iret;

  char dummy[MAXLINE];
  int n = MAXLINE;

  if (NULL == (fp = fopen(filename,"r")))
  {
    fprintf(stderr, "pgmsize: cannot open <%s>\n", filename);
    exit(-1);
  }

  cret = fgets(dummy, n, fp);
  cret = fgets(dummy, n, fp);

  iret = fscanf(fp,"%d %d", nx, ny);
      
  fclose(fp);
}


/*
 *  Routine to read a PGM data file into a 2D floating point array
 *  x[nx][ny]. Because of the way C handles (or fails to handle!)
 *  multi-dimensional arrays we have to cast the pointer to void.
 *
 *  Note that this assumes a single line comment and no other white space.
 */

void pgmread(char *filename, void *vx, int nx, int ny)
{ 
  FILE *fp;

  int nxt, nyt, i, j, t;
  char dummy[MAXLINE];
  int n = MAXLINE;

  char *cret;
  int iret;

  float *x = (float *) vx;

  if (NULL == (fp = fopen(filename,"r")))
  {
    fprintf(stderr, "pgmread: cannot open <%s>\n", filename);
    exit(-1);
  }

  cret = fgets(dummy, n, fp);
  cret = fgets(dummy, n, fp);

  iret = fscanf(fp,"%d %d",&nxt,&nyt);

  if (nx != nxt || ny != nyt)
  {
    fprintf(stderr,
            "pgmread: size mismatch, (nx,ny) = (%d,%d) expected (%d,%d)\n",
            nxt, nyt, nx, ny);
    exit(-1);
  }

  iret = fscanf(fp,"%d",&i);

  /*
   *  Must cope with the fact that the storage order of the data file
   *  is not the same as the storage of a C array, hence the pointer
   *  arithmetic to access x[i][j].
   */

  for (j=0; j<ny; j++)
  {
    for (i=0; i<nx; i++)
    {
      iret = fscanf(fp,"%d", &t);
      x[(ny-j-1)+ny*i] = t;
    }
  }

  fclose(fp);
}


/*
 *  Routine to write a PGM image file from a 2D floating point array
 *  x[nx][ny]. Because of the way C handles (or fails to handle!)
 *  multi-dimensional arrays we have to cast the pointer to void.
 */

void pgmwrite(char *filename, void *vx, int nx, int ny)
{
  FILE *fp;

  int i, j, k, grey;

  float xmin, xmax, tmp, fval;
  float thresh = 255.0;

  float *x = (float *) vx;

  if (NULL == (fp = fopen(filename,"w")))
  {
    fprintf(stderr, "pgmwrite: cannot create <%s>\n", filename);
    exit(-1);
  }

  printf("Writing %d x %d picture into file: %s\n", nx, ny, filename);

  /*
   *  Find the max and min absolute values of the array
   */

  xmin = fabs(x[0]);
  xmax = fabs(x[0]);

  for (i=0; i < nx*ny; i++)
  {
    if (fabs(x[i]) < xmin) xmin = fabs(x[i]);
    if (fabs(x[i]) > xmax) xmax = fabs(x[i]);
  }

  if (xmin == xmax) xmin = xmax-1.0;

  fprintf(fp, "P2\n");
  fprintf(fp, "# Written by pgmio::pgmwrite\n");
  fprintf(fp, "%d %d\n", nx, ny);
  fprintf(fp, "%d\n", (int) thresh);

  k = 0;

  for (j=ny-1; j >=0 ; j--)
  {
    for (i=0; i < nx; i++)
    {
      /*
       *  Access the value of x[i][j]
       */

      tmp = x[j+ny*i];

      /*
       *  Scale the value appropriately so it lies between 0 and thresh
       */

      fval = thresh*((fabs(tmp)-xmin)/(xmax-xmin))+0.5;
      grey = (int) fval;

      fprintf(fp, "%3d ", grey);

      if (0 == (k+1)%16) fprintf(fp, "\n");

      k++;
    }
  }

  if (0 != k%16) fprintf(fp, "\n");
  fclose(fp);
}

File: MPP-casestudy/pgmio.f90
!
!  This file contains Fortran routines for the MPI Casestudy.
!
!  "pgmread" reads in a PGM picture and can be called as follows:
!
!     real buf(M,N)
!     call pgmread('edge.pgm', buf)
!
!  "pgmwrite" writes an array as PGM picture and can be called as follows:
!
!     real buf(M,N)
!     call pgmwrite('picture.pgm', buf)
!
!  "pgmsize" returns the size of a PGM picture and can be called as follows:
!
!     integer :: nx, ny
!     call pgmsize('edge.pgm', nx, ny)
!
!  To access these routines, add the following to your program:
!
!     use pgmio
!

module pgmio

implicit none

integer, parameter, private :: iounit = 10
integer, parameter, private :: realnumber=kind(1.0e0)

real(kind=realnumber), parameter, private :: thresh = 255.0


contains


!  Routine to get the size of a PGM image
!  Note that this assumes a single line comment and no other white space

subroutine pgmsize(filename, nx, ny)

  integer :: nx, ny

  character*(*) :: filename

  open(unit=iounit, file=filename)

  read(iounit,*)
  read(iounit,*)
  read(iounit,*) nx, ny

  close(unit=iounit)

end subroutine pgmsize


!  Routine to read a PGM file into a 2D floating-point array
!  Note that this assumes a single line comment and no other white space


subroutine pgmread(filename, x)

  character*(*) :: filename
  integer :: nx, ny, nxt, nyt
  real(kind=realnumber), dimension(:,:) :: x

  integer i, j

  nx = size(x,1)
  ny = size(x,2)

  write(*,*) 'Reading ', nx, ' x ', ny, ' picture from: ', filename

  open(unit=iounit, file=filename)

  read(iounit,*)
  read(iounit,*)
  read(iounit,*) nxt, nyt

  if (nx .ne. nxt .or. ny .ne. nyt) then
    write(*,*) 'pgmread: size mismatch, (nx,ny) = (', nxt, ',', nyt, &
               ') expected (', nx, ',', ny, ')'
    stop
  end if

  read(iounit,*)
  read(iounit,*) ((x(i,ny-j+1), i=1,nx), j=1,ny)

  close(unit=iounit)

end subroutine pgmread


!  Routine to write a PGM image file from a 2D floating-point array
!  To cope with negative numbers we take the absolute value
!  Output values are scaled to lie between 0 and thresh

subroutine pgmwrite(filename, x)

  implicit none

  character*(*) :: filename
  integer :: nx, ny

  real(kind=realnumber), dimension(:,:) :: x

  real(kind=realnumber) :: tmin, tmax

  integer :: i, j

  nx = size(x,1)
  ny = size(x,2)

  write(*,*) 'Writing ', nx, ' x ', ny, ' picture into: ', filename

  open(unit=iounit, file=filename)

  tmin = minval(abs(x(:,:)))
  tmax = maxval(abs(x(:,:)))

  if (tmin == tmax) tmin = tmax-1

  write(iounit,fmt='(''P2''/''# Written by pgmio::pgmwrite'')')
  write(iounit,fmt='(i4, '' '', i4)') nx, ny

  write(iounit,*) int(thresh)
  write(iounit,fmt='(18(i3,'' ''))') &
       ((int(thresh*(abs(x(i,ny-j+1))-tmin)/(tmax-tmin)+0.5), i=1,nx), j=1,ny)

  close(unit=iounit)

end subroutine pgmwrite

end module pgmio
